File: apps/backend/src/services/object-storage.ts
import fs from 'fs/promises'
import fsSync from 'fs'
import path from 'path'
import {
  DeleteObjectCommand,
  PutObjectCommand,
  S3Client,
} from '@aws-sdk/client-s3'

type S3BackendConfig = {
  name: string
  endpoint?: string
  bucket: string
  region?: string
  accessKey?: string
  secretKey?: string
  publicBaseUrl?: string
  forcePathStyle?: boolean
}

type S3Backend = {
  name: string
  client: S3Client
  bucket: string
  publicBaseUrl: string
}

type UploadBufferOptions = {
  contentType?: string
  s3Key?: string
}

function trimTrailingSlash(value: string) {
  return value.replace(/\/+$/, '')
}

function sanitizeKeyPart(value: string) {
  return value.replace(/[^a-zA-Z0-9._-]+/g, '_')
}

function toBuffer(value: Buffer | Uint8Array | ArrayBuffer) {
  if (Buffer.isBuffer(value)) {
    return value
  }

  if (value instanceof ArrayBuffer) {
    return Buffer.from(value)
  }

  return Buffer.from(value)
}

function parseS3Backends(): S3Backend[] {
  const raw = process.env.S3_BACKENDS?.trim()
  if (!raw) return []

  let configs: S3BackendConfig[]
  try {
    configs = JSON.parse(raw)
    if (!Array.isArray(configs) || configs.length === 0) return []
  } catch {
    console.warn('[ObjectStorage] 无法解析 S3_BACKENDS，回退到传统单后端配置')
    return []
  }

  const backends: S3Backend[] = []
  for (const cfg of configs) {
    if (!cfg.name || !cfg.bucket) {
      console.warn('[ObjectStorage] 跳过无效的后端配置:', cfg)
      continue
    }

    backends.push({
      name: cfg.name,
      client: new S3Client({
        region: cfg.region || 'auto',
        forcePathStyle: cfg.forcePathStyle ?? true,
        ...(cfg.endpoint ? { endpoint: cfg.endpoint } : {}),
        ...(cfg.accessKey && cfg.secretKey
          ? {
              credentials: {
                accessKeyId: cfg.accessKey,
                secretAccessKey: cfg.secretKey,
              },
            }
          : {}),
      }),
      bucket: cfg.bucket,
      publicBaseUrl: trimTrailingSlash(cfg.publicBaseUrl || ''),
    })
  }

  return backends
}

function buildLegacyBackend(): S3Backend | null {
  const provider = (process.env.STORAGE_PROVIDER ?? 'local').toLowerCase()
  if (provider !== 's3' && provider !== 'r2') return null

  const bucket = process.env.S3_BUCKET ?? ''
  if (!bucket) {
    throw new Error('Missing S3_BUCKET environment variable')
  }

  return {
    name: provider,
    client: new S3Client({
      region: process.env.S3_REGION?.trim() || 'auto',
      forcePathStyle: provider !== 'r2',
      ...(process.env.S3_ENDPOINT?.trim()
        ? { endpoint: process.env.S3_ENDPOINT.trim() }
        : {}),
      ...(process.env.S3_ACCESS_KEY_ID?.trim() &&
      process.env.S3_SECRET_ACCESS_KEY?.trim()
        ? {
            credentials: {
              accessKeyId: process.env.S3_ACCESS_KEY_ID.trim(),
              secretAccessKey: process.env.S3_SECRET_ACCESS_KEY.trim(),
            },
          }
        : {}),
    }),
    bucket,
    publicBaseUrl: trimTrailingSlash(process.env.S3_PUBLIC_BASE_URL?.trim() || ''),
  }
}

export class ObjectStorageService {
  private readonly uploadDir = path.resolve(
    process.cwd(),
    process.env.UPLOAD_DIR ?? './uploads',
  )
  private readonly backends: S3Backend[]

  constructor() {
    const parsed = parseS3Backends()
    if (parsed.length > 0) {
      this.backends = parsed
      console.log(
        `[ObjectStorage] 已配置 ${this.backends.length} 个 S3 后端:`,
        this.backends.map((b) => b.name).join(', '),
      )
      return
    }

    const legacy = buildLegacyBackend()
    if (legacy) {
      this.backends = [legacy]
      return
    }

    this.backends = []
  }

  getProvider() {
    if (this.backends.length > 1) return 'multi'
    if (this.backends.length === 1) return this.backends[0]!.name
    return 'local'
  }

  getLocalPath(key: string) {
    return path.join(this.uploadDir, key)
  }

  getPublicLocation(key: string) {
    if (this.backends.length > 0 && this.backends[0]!.publicBaseUrl) {
      return `${this.backends[0]!.publicBaseUrl}/${key}`
    }

    if (this.backends.length > 0) {
      return `s3://${this.backends[0]!.bucket}/${key}`
    }

    return `local://${key}`
  }

  async uploadBuffer(
    key: string,
    buffer: Buffer | Uint8Array | ArrayBuffer,
    options: UploadBufferOptions = {},
  ) {
    const body = toBuffer(buffer)
    const localPath = this.getLocalPath(key)

    await fs.mkdir(path.dirname(localPath), { recursive: true })
    await fs.writeFile(localPath, body)

    if (this.backends.length > 0) {
      const remoteKey = options.s3Key ?? key
      const results = await Promise.allSettled(
        this.backends.map((backend) =>
          backend.client.send(
            new PutObjectCommand({
              Bucket: backend.bucket,
              Key: remoteKey,
              Body: body,
              ContentType: options.contentType ?? 'application/octet-stream',
            }),
          ),
        ),
      )

      for (let i = 0; i < results.length; i++) {
        const result = results[i]
        if (result?.status === 'rejected') {
          console.warn(
            `[ObjectStorage] 上传到 ${this.backends[i]!.name} 失败:`,
            result.reason,
          )
        }
      }
    }

    return {
      key,
      localPath,
      publicLocation: this.getPublicLocation(options.s3Key ?? key),
      size: body.length,
    }
  }

  async deleteObject(key: string, s3Key?: string) {
    const localPath = this.getLocalPath(key)
    await fs.unlink(localPath).catch(() => undefined)

    if (this.backends.length > 0) {
      const remoteKey = s3Key ?? key
      const results = await Promise.allSettled(
        this.backends.map((backend) =>
          backend.client.send(
            new DeleteObjectCommand({
              Bucket: backend.bucket,
              Key: remoteKey,
            }),
          ),
        ),
      )

      for (let i = 0; i < results.length; i++) {
        const result = results[i]
        if (result?.status === 'rejected') {
          console.warn(
            `[ObjectStorage] 从 ${this.backends[i]!.name} 删除失败:`,
            result.reason,
          )
        }
      }
    }
  }

  async exists(key: string) {
    const localPath = this.getLocalPath(key)
    return fsSync.existsSync(localPath)
  }

  sanitizeKeyPart(value: string) {
    return sanitizeKeyPart(value)
  }
}

export const storageService = new ObjectStorageService()

File: apps/backend-rs/src/services/object_storage.rs
use aws_config::BehaviorVersion;
use aws_sdk_s3::Client as S3Client;
use aws_sdk_s3::config::Region;
use aws_sdk_s3::operation::put_object::PutObjectOutput;
use aws_sdk_s3::primitives::ByteStream;
use serde::Deserialize;
use std::path::Path;

use crate::config::Config;

#[derive(Debug, Clone, Deserialize)]
pub struct S3BackendConfig {
    pub name: String,
    pub endpoint: Option<String>,
    pub bucket: String,
    pub region: Option<String>,
    pub access_key: Option<String>,
    pub secret_key: Option<String>,
    pub public_base_url: Option<String>,
    #[serde(default = "default_true")]
    pub force_path_style: bool,
}

fn default_true() -> bool {
    true
}

pub struct S3Backend {
    pub name: String,
    client: S3Client,
    pub bucket: String,
    pub public_base_url: String,
}

pub struct ObjectStorageService {
    pub backends: Vec<S3Backend>,
    pub upload_dir: String, // local FS fallback
}

impl ObjectStorageService {
    /// Build from config, supporting both S3_BACKENDS JSON and legacy env vars.
    pub async fn new(cfg: &Config) -> Self {
        let mut backends = Vec::new();

        // Parse S3_BACKENDS JSON array
        if let Some(json) = &cfg.s3_backends_json
            && let Ok(configs) = serde_json::from_str::<Vec<S3BackendConfig>>(json)
        {
            for bc in configs {
                if let Some(s3) = Self::build_backend(&bc).await {
                    backends.push(s3);
                }
            }
        }

        // Parse legacy env vars
        if backends.is_empty() && !cfg.s3_bucket.is_empty() {
            let legacy = S3BackendConfig {
                name: "legacy".into(),
                endpoint: cfg.s3_endpoint.clone(),
                bucket: cfg.s3_bucket.clone(),
                region: Some(cfg.s3_region.clone()),
                access_key: cfg.s3_access_key_id.clone(),
                secret_key: cfg.s3_secret_access_key.clone(),
                public_base_url: cfg.s3_public_base_url.clone(),
                force_path_style: true,
            };
            if let Some(s3) = Self::build_backend(&legacy).await {
                backends.push(s3);
            }
        }

        Self {
            backends,
            upload_dir: cfg.upload_dir.clone(),
        }
    }

    async fn build_backend(bc: &S3BackendConfig) -> Option<S3Backend> {
        let region = bc.region.clone().unwrap_or_else(|| "auto".to_string());

        let mut loader =
            aws_config::defaults(BehaviorVersion::latest()).region(Region::new(region.clone()));

        // Custom endpoint + credentials
        if let (Some(ak), Some(sk)) = (&bc.access_key, &bc.secret_key) {
            use aws_sdk_s3::config::Credentials;
            let creds = Credentials::new(ak, sk, None, None, "env");
            loader = loader.credentials_provider(creds);
        }

        let sdk_config = loader.load().await;

        let mut s3_config =
            aws_sdk_s3::config::Builder::from(&sdk_config).force_path_style(bc.force_path_style);

        if let Some(endpoint) = &bc.endpoint {
            s3_config = s3_config.endpoint_url(endpoint).region(Region::new(region));
        }

        let client = S3Client::from_conf(s3_config.build());

        Some(S3Backend {
            name: bc.name.clone(),
            client,
            bucket: bc.bucket.clone(),
            public_base_url: bc
                .public_base_url
                .clone()
                .unwrap_or_default()
                .trim_end_matches('/')
                .to_string(),
        })
    }

    /// Sanitize path segment for object keys.
    pub fn sanitize_key_part(s: &str) -> String {
        s.chars()
            .map(|c| {
                if c.is_alphanumeric() || c == '-' || c == '_' || c == '.' {
                    c
                } else {
                    '_'
                }
            })
            .collect()
    }

    /// Local FS path for a key.
    pub fn local_path(&self, key: &str) -> String {
        Path::new(&self.upload_dir)
            .join(key)
            .to_string_lossy()
            .to_string()
    }

    /// Check local file existence.
    pub fn local_exists(&self, key: &str) -> bool {
        Path::new(&self.local_path(key)).exists()
    }

    /// Upload buffer to local FS and all S3 backends.
    pub async fn upload(&self, key: &str, data: &[u8]) -> Result<Vec<PutObjectOutput>, String> {
        // Write to local FS
        let local_path = self.local_path(key);
        if let Some(parent) = Path::new(&local_path).parent() {
            tokio::fs::create_dir_all(parent)
                .await
                .map_err(|e| format!("mkdir failed: {e}"))?;
        }
        tokio::fs::write(&local_path, data)
            .await
            .map_err(|e| format!("write failed: {e}"))?;

        // Upload to each S3 backend
        let mut results = Vec::new();
        for backend in &self.backends {
            let body = ByteStream::from(data.to_vec());
            match backend
                .client
                .put_object()
                .bucket(&backend.bucket)
                .key(key)
                .body(body)
                .send()
                .await
            {
                Ok(out) => results.push(out),
                Err(e) => {
                    tracing::warn!("S3 upload to {} failed for key {key}: {e}", backend.name);
                }
            }
        }
        Ok(results)
    }

    /// Delete from local FS and all S3 backends.
    pub async fn delete(&self, key: &str) {
        // Local
        let local_path = self.local_path(key);
        let _ = tokio::fs::remove_file(&local_path).await;

        // S3
        for backend in &self.backends {
            if let Err(e) = backend
                .client
                .delete_object()
                .bucket(&backend.bucket)
                .key(key)
                .send()
                .await
            {
                tracing::warn!("S3 delete on {} failed for key {key}: {e}", backend.name);
            }
        }
    }

    /// Get public URL for a key (first backend or local path).
    pub fn public_url(&self, key: &str) -> String {
        if let Some(backend) = self.backends.first()
            && !backend.public_base_url.is_empty()
        {
            return format!("{}/{}", backend.public_base_url, key);
        }
        self.local_path(key)
    }

    /// Check if we have any S3 backends configured.
    pub fn has_s3(&self) -> bool {
        !self.backends.is_empty()
    }
}
